File: remote/src/main/java/org/openbase/bco/dal/remote/trigger/preset/NeighborConnectionPresenceTrigger.java
package org.openbase.bco.dal.remote.trigger.preset;

/*-
 * #%L
 * BCO DAL Remote
 * %%
 * Copyright (C) 2014 - 2017 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */
import org.openbase.bco.dal.remote.unit.connection.ConnectionRemote;
import org.openbase.bco.dal.remote.unit.location.LocationRemote;
import org.openbase.jul.pattern.trigger.AbstractTrigger;
import org.openbase.jul.exception.CouldNotPerformException;
import org.openbase.jul.exception.printer.ExceptionPrinter;
import org.openbase.jul.extension.rst.processing.TimestampProcessor;
import org.openbase.jul.pattern.Observable;
import org.openbase.jul.pattern.Observer;
import org.openbase.jul.pattern.Remote;
import org.slf4j.LoggerFactory;
import rst.domotic.state.ActivationStateType.ActivationState;
import rst.domotic.state.DoorStateType.DoorState;
import rst.domotic.state.PresenceStateType.PresenceState;
import rst.domotic.state.WindowStateType.WindowState;

/**
 *
 * @author <a href="mailto:tmichalski@techfak.uni-bielefeld.de">Timo Michalski</a>
 */
public class NeighborConnectionPresenceTrigger extends AbstractTrigger {

    private final Observer dataObserver;
    private final Observer<Remote.ConnectionState> connectionObserver;
    private final LocationRemote locationRemote;
    private final ConnectionRemote connectionRemote;
    private boolean active = false;

    public NeighborConnectionPresenceTrigger(final LocationRemote locationRemote, final ConnectionRemote connectionRemote) throws org.openbase.jul.exception.InstantiationException {
        super();

        this.locationRemote = locationRemote;
        this.connectionRemote = connectionRemote;

        dataObserver = (source, data) -> {
            verifyCondition();
        };

        connectionObserver = (Observable<Remote.ConnectionState> source, Remote.ConnectionState data) -> {
            if (data.equals(Remote.ConnectionState.CONNECTED)) {
                verifyCondition();
            } else {
                notifyChange(TimestampProcessor.updateTimestampWithCurrentTime(ActivationState.newBuilder().setValue(ActivationState.State.UNKNOWN).build()));
            }
        };
    }

    @Override
    public void activate() throws CouldNotPerformException, InterruptedException {
        locationRemote.addDataObserver(dataObserver);
        connectionRemote.addDataObserver(dataObserver);
        locationRemote.addConnectionStateObserver(connectionObserver);
        connectionRemote.addConnectionStateObserver(connectionObserver);
        active = true;
        verifyCondition();
    }

    @Override
    public void deactivate() throws CouldNotPerformException, InterruptedException {
        locationRemote.removeDataObserver(dataObserver);
        connectionRemote.removeDataObserver(dataObserver);
        locationRemote.removeConnectionStateObserver(connectionObserver);
        connectionRemote.removeConnectionStateObserver(connectionObserver);
        active = false;
        notifyChange(TimestampProcessor.updateTimestampWithCurrentTime(ActivationState.newBuilder().setValue(ActivationState.State.UNKNOWN).build()));
    }

    @Override
    public boolean isActive() {
        return active;
    }

    @Override
    public void shutdown() {
        try {
            deactivate();
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
        } catch (CouldNotPerformException ex) {
            ExceptionPrinter.printHistory("Could not shutdown " + this, ex, LoggerFactory.getLogger(getClass()));
        }
        super.shutdown();
    }

    private void verifyCondition() {
        try {
            if (locationRemote.getData().getPresenceState().getValue().equals(PresenceState.State.PRESENT)
                    && (connectionRemote.getDoorState().getValue().equals(DoorState.State.OPEN)
                    || connectionRemote.getWindowState().getValue().equals(WindowState.State.OPEN))) {
                notifyChange(TimestampProcessor.updateTimestampWithCurrentTime(ActivationState.newBuilder().setValue(ActivationState.State.ACTIVE).build()));
            } else {
                notifyChange(TimestampProcessor.updateTimestampWithCurrentTime(ActivationState.newBuilder().setValue(ActivationState.State.DEACTIVE).build()));
            }
        } catch (CouldNotPerformException ex) {
            ExceptionPrinter.printHistory("Could not verify trigger state " + this, ex, LoggerFactory.getLogger(getClass()));
        }
    }
}

File: module/dal/remote/src/main/java/org/openbase/bco/dal/remote/trigger/preset/NeighborConnectionPresenceTrigger.kt
package org.openbase.bco.dal.remote.trigger.preset

import org.openbase.bco.dal.remote.layer.unit.connection.ConnectionRemote
import org.openbase.bco.dal.remote.layer.unit.location.LocationRemote
import org.openbase.jul.exception.CouldNotPerformException
import org.openbase.jul.exception.printer.ExceptionPrinter
import org.openbase.jul.extension.type.processing.TimestampProcessor
import org.openbase.jul.pattern.Observer
import org.openbase.jul.pattern.controller.Remote
import org.openbase.jul.pattern.provider.DataProvider
import org.openbase.jul.pattern.trigger.AbstractTrigger
import org.openbase.type.domotic.state.*
import org.openbase.type.domotic.unit.connection.ConnectionDataType
import org.openbase.type.domotic.unit.location.LocationDataType
import org.slf4j.LoggerFactory

/*-
 * #%L
 * BCO DAL Remote
 * %%
 * Copyright (C) 2014 - 2021 openbase.org
 * %%
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Lesser Public License for more details.
 * 
 * You should have received a copy of the GNU General Lesser Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/lgpl-3.0.html>.
 * #L%
 */ /**
 *
 * @author [Timo Michalski](mailto:tmichalski@techfak.uni-bielefeld.de)
 */
class NeighborConnectionPresenceTrigger(
    private val locationRemote: LocationRemote,
    private val connectionRemote: ConnectionRemote
) : AbstractTrigger() {
    private val locationObserver: Observer<DataProvider<LocationDataType.LocationData>, LocationDataType.LocationData>
    private val connectionObserver: Observer<DataProvider<ConnectionDataType.ConnectionData>, ConnectionDataType.ConnectionData>
    private val connectionStateObserver: Observer<Remote<*>, ConnectionStateType.ConnectionState.State>
    private var active = false

    init {
        locationObserver = Observer { _, _ -> verifyCondition() }
        connectionObserver = Observer {  _, _ -> verifyCondition() }
        connectionStateObserver = Observer { _, data: ConnectionStateType.ConnectionState.State ->
            if (data == ConnectionStateType.ConnectionState.State.CONNECTED) {
                verifyCondition()
            } else {
                notifyChange(
                    TimestampProcessor.updateTimestampWithCurrentTime(
                        ActivationStateType.ActivationState.newBuilder().setValue(
                            ActivationStateType.ActivationState.State.UNKNOWN
                        ).build()
                    )
                )
            }
        }
    }

    @Throws(CouldNotPerformException::class, InterruptedException::class)
    override fun activate() {
        locationRemote.addDataObserver(locationObserver)
        connectionRemote.addDataObserver(connectionObserver)
        locationRemote.addConnectionStateObserver(connectionStateObserver)
        connectionRemote.addConnectionStateObserver(connectionStateObserver)
        active = true
        verifyCondition()
    }

    @Throws(CouldNotPerformException::class, InterruptedException::class)
    override fun deactivate() {
        locationRemote.removeDataObserver(locationObserver)
        connectionRemote.removeDataObserver(connectionObserver)
        locationRemote.removeConnectionStateObserver(connectionStateObserver)
        connectionRemote.removeConnectionStateObserver(connectionStateObserver)
        active = false
        notifyChange(
            TimestampProcessor.updateTimestampWithCurrentTime(
                ActivationStateType.ActivationState.newBuilder().setValue(
                    ActivationStateType.ActivationState.State.UNKNOWN
                ).build()
            )
        )
    }

    override fun isActive(): Boolean {
        return active
    }

    override fun shutdown() {
        try {
            deactivate()
        } catch (ex: InterruptedException) {
            Thread.currentThread().interrupt()
        } catch (ex: CouldNotPerformException) {
            ExceptionPrinter.printHistory("Could not shutdown $this", ex, LoggerFactory.getLogger(javaClass))
        }
        super.shutdown()
    }

    private fun verifyCondition() {
        try {
            if (locationRemote.data.presenceState.value == PresenceStateType.PresenceState.State.PRESENT && connectionRemote.doorState.value == DoorStateType.DoorState.State.OPEN || connectionRemote.windowState.value == WindowStateType.WindowState.State.OPEN) {
                notifyChange(
                    TimestampProcessor.updateTimestampWithCurrentTime(
                        ActivationStateType.ActivationState.newBuilder().setValue(
                            ActivationStateType.ActivationState.State.ACTIVE
                        ).build()
                    )
                )
            } else {
                notifyChange(
                    TimestampProcessor.updateTimestampWithCurrentTime(
                        ActivationStateType.ActivationState.newBuilder().setValue(
                            ActivationStateType.ActivationState.State.INACTIVE
                        ).build()
                    )
                )
            }
        } catch (ex: CouldNotPerformException) {
            ExceptionPrinter.printHistory(
                "Could not verify trigger state $this",
                ex,
                LoggerFactory.getLogger(javaClass)
            )
        }
    }
}
